replica/database.cc

```diff
@@ -899,6 +918,15 @@ sstables::sstables_manager& database::get_sstables_manager(const schema& s) cons
     return get_sstables_manager(system_keyspace(is_system_table(s)));
 }
 
+static bool shall_reject_write(const database& db, const schema& s)
+{
+    // MVs/SIs, CDC Log table - are located in the same keyspace as the associated user table
+    // audit - is not an internal keyspace in the sens of data_dictionary::keyspace::is_internal
+    // cql tracing - is part of system_traces that is an internal keyspace but we want to reject
+    return db.is_in_critical_disk_utilization_mode() &&
+        (!db.as_data_dictionary().find_keyspace(s.ks_name()).is_internal() || s.ks_name() == tracing::trace_keyspace_helper::KEYSPACE_NAME);
```

Comment:
find_keyspace searches the keyspace name and this is done on the hot write path, for every write.
Since `find_column_family` is called anyhow later on, we can call it earlier and tag the tables in `database::tables_metadata::add_table_helper` based on a predicate for keyspaces (for `is_internal() || s.ks_name() == tracing::trace_keyspace_helper::KEYSPACE_NAME`) so we can have a `tbl.shall_reject_writes_in_critical_disk_utilization_mode()` predicate on the table itself.

Reply:
A back-pointer from the table to the keyspace object could be useful too, but I'm not sure how easy it would be to implement one.
Alternatively, we can also maintain something like `std::unordered_map<table_id, lw_shared_ptr<keyspace>> replica::database::_keyspace_by_table_id` and implement `std::tuple<table&, keyspace&> find_table_and_keyspace(table_id)` that will return both, but it will cost another unordered_map search,

Reply:
I will have a look at it.

Reply:
I tried to follow the second approach that I liked more, but the change became quite intrusive. Eventually, after fighting with issues, I turned towards the first suggestion and got implemented.